src/main/kotlin/com/criptext/mail/utils/generaldatasource/data/UserDataWriter.kt

```diff
@@ -21,22 +19,157 @@ class UserDataWriter(private val emailDao: EmailDao,
 
             val tmpFileLinkData = createTempFile()
 
-            addContactsToFile(contactDao.getAll(), tmpFileLinkData)
-            addLabelsToFile(labelDao.getAll(), tmpFileLinkData)
-            addFilesToFile(fileDao.getAll(), tmpFileLinkData)
-            addMailsToFile(emailDao.getAll(), tmpFileLinkData)
-            addMailsAndLabelsRelationToFile(emailLabelDao.getAll(), tmpFileLinkData)
-            addMailsAndContactsRelationToFile(emailContactDao.getAll(), tmpFileLinkData)
-            addMailsFileKey(fileKeyDao.getAll(), tmpFileLinkData)
-            addExternalSessionsToFile(emailExternalSessionDao.getAll(), tmpFileLinkData)
-
+            addContactsToFile(db.contactDao().getAll(), tmpFileLinkData)
+            addLabelsToFile(db.labelDao().getAll(), tmpFileLinkData)
+            addMailsToFile(db.emailDao().getAllForLinkFile(), tmpFileLinkData)
+            addFilesToFile(db.fileDao().getAllForLinkFile(), tmpFileLinkData)
+            addMailsAndLabelsRelationToFile(db.emailLabelDao().getAllForLinkFile(), tmpFileLinkData)
+            addMailsAndContactsRelationToFile(db.emailContactDao().getAllForLinkFile(), tmpFileLinkData)
+            addMailsFileKey(db.fileKeyDao().getAll(), tmpFileLinkData)
 
             return tmpFileLinkData.absolutePath
         }catch (ex:Exception){
             return null
         }
     }
 
+    fun createDBFromFile(file: File) {
+        val contactList = mutableListOf<Contact>()
+        val labelList = mutableListOf<Label>()
+        val fileList = mutableListOf<CRFile>()
+        val emailList = mutableListOf<Email>()
+        val emailLabelList = mutableListOf<EmailLabel>()
+        val emailContactList = mutableListOf<EmailContact>()
+        val fileKeyList = mutableListOf<FileKey>()
+        val data = file.bufferedReader()
+        var line = data.readLine()
+        db.beginTransaction()
+        while(line != null) {
+            val json = JSONObject(line)
+            when (json.getString("table")) {
+                "contact" -> {
+                    val contact = Contact.fromJSON(json.get("object").toString())
+                    if(contactList.size == DEFAULT_BATCH_SIZE){
+                        db.contactDao().insertAll(contactList)
+                        contactList.clear()
+                    }
+                    contactList.add(contact)
```

Comment:
This is too repetitive. This logic should be moved to a class `BackupDataWriter`, something like this:

``` kotlin
abstract class BackupDataWriter<T> {
    abstract fun deserialize(item: String)
    abstract fun writeBatch(batch: List<T>)
    // calls deserialize and and adds item to batch. inserts to db when batch is ready
    fun insert(item: String) 
    fun flush()
}
```

then you could have something like: 

``` kotlin
when (json.getString("table") {
    "contact" -> contactWriter.insert(json.get("object").toString())
    "label" -> labelWriter.insert(json.get("object").toString())
    "email" -> emailWriter.insert(json.get("object").toString())
    // ....
}
```